### ISSUE TITLE
Binance Futures Invalid API-key, IP, or permissions for action

### ISSUE THREAD

**Issue by cadmufasa** [ccxt/ccxt#7263]
Hi,

I am trying to connect to the Futures servers on Binance (and access the contents in the Futures Wallet)

I had created the API key several times on Binance.

I have followed the suggestions listed here and created a NEW API Key - several times:
https://dev.binance.vision/t/why-do-i-see-this-error-invalid-api-key-ip-or-permissions-for-action/93
https://binance.zendesk.com/hc/en-us/articles/360043210712-Updates-to-the-Binance-Futures-API

When attempting to connect, I get:
MESSAGE : ERR-52D073683{"code":-2015,"msg":"Invalid API-key, IP, or permissions for action, request ip: 144.202.27.147"}

So, I CHANGED the WORKING API KEY so that it would process FUTUREs request too. I REMOVED the NEW key

I still got the same error:

When attempting to connect, I get:
MESSAGE : ERR-52D073683{"code":-2015,"msg":"Invalid API-key, IP, or permissions for action, request ip: 144.202.27.147"}

I used to have a white-listed IP Address but I have removed that option (in an effort to figrue out what is going wrong).

I was not sure if it had something to do with the API servers being used. 

What other things can I try to figure out what the problem is. Why is it taking place?

Any help,  hints or advice would be greatly appreciated.

Regards.

I have the following code below :

---------------------------------------------
-- initialize the connection
---------------------------------------------
def initialize_connection( in_futures = False):

    try :

        if ( in_futures == False ):
            LD_API_KEY      = os.getenv("BINANCE_APIKEY")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")

            conn_exchange = ccxt.binance({
                'apiKey'         : LD_API_KEY,
                'secret'         : LD_API_SECRET,
                'enableRateLimit': True,
            })

        elif ( in_futures == True ):
  
            LD_API_KEY      = os.getenv("BINANCE_APIKEY_MARGIN")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET_MARGIN")

            conn_exchange = ccxt.binance({
                'apiKey': LD_API_KEY,
                'secret': LD_API_SECRET,
                'timeout': 30000,
                'enableRateLimit': True,
                'options': { 'defaultType': 'future', },
            })

        else:
            raise CustomProgramError('11408', 'invalid choice made ' + str( in_futures) )

        return( conn_exchange )

---------------------------------------------
-- getting the balances:
---------------------------------------------
def get_bank_balances(in_futures = False, in_symbol = '', print_info = False):
    try :

	[ ... snip ...]

        exchange = initialize_connection( in_futures )  

	[ ... snip ...]

        # reference: https://unpkg.com/ccxt@1.10.633/doc/manual.rst
        wallets            = exchange.fetch_balance()

	[ ... snip ...]

---------------------------------------------
-- error message
---------------------------------------------

2020-07-13 14:17:43 - ERROR - 26699 - test - ERROR
MESSAGE : ERR-9E25D696D{"code":-2015,"msg":"Invalid API-key, IP, or permissions for action, request ip: 144.202.27.147"}

2020-07-13 14:17:43 - ERROR - 26699 - test - ERROR
MESSAGE : ERR-9E25D696D - Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/ccxt/base/exchange.py", line 568, in fetch
    response.raise_for_status()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://fapi.binance.com/fapi/v1/account?timestamp=1594649863166&recvWindow=5000&signature=39503a217abc23o2o4075fc3d03af19c2c2b56372b018d3940693fk422c2782

**Comment by kroitor** [comment]
@cadmufasa 

1. Binance API keys for futures are the same as for SPOT.
2. Before opening a new issue make sure you've followed the troubleshooting guides:

  - https://github.com/ccxt/ccxt/wiki/Manual#troubleshooting

3. Then always try the search first:

  - https://github.com/ccxt/ccxt/search?q=binance+Invalid+API-key&type=Issues
  - https://github.com/ccxt/ccxt/issues/6699
  - https://github.com/ccxt/ccxt/issues/3057
  - https://github.com/ccxt/ccxt/issues/2044
  - https://github.com/ccxt/ccxt/issues/320

4. And always paste your versions and verbose output precisely as explained here:

  - https://github.com/ccxt/ccxt/wiki/FAQ#what-is-required-to-get-help

> What other things can I try to figure out what the problem is. Why is it taking place?

2. Make sure that your key is correct. Also, make sure that you didn't misplace apiKey←→secret. 
You can check which key is being sent in verbose mode:

```Python
import ccxt
import os
from pprint import pprint

print('CCXT Version:', ccxt.__version__)

LD_API_KEY = os.getenv("BINANCE_APIKEY_MARGIN")
LD_API_SECRET = os.getenv("BINANCE_API_SECRET_MARGIN")

conn_exchange = ccxt.binance({
    'apiKey': LD_API_KEY,
    'secret': LD_API_SECRET,
    'enableRateLimit': True,
    'options': { 'defaultType': 'future' },
})

exchange.load_markets()

exchange.verbose = True
pprint(exchange.fetch_balance())
```

↑ Print your complete verbose output (request + response) here.

5. Make sure you have enabled the futures trading on Binance's website. Does placing orders work for you with the website?

**Comment by cadmufasa** [comment]
Thanks for the response.

I use the following with the current API KEY (and the Futures option ticked)

            LD_API_KEY      = os.getenv("BINANCE_APIKEY")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")

            conn_exchange = ccxt.binance({
                'apiKey'         : LD_API_KEY,
                'secret'         : LD_API_SECRET,
                'enableRateLimit': True,
            })

This Works (of course this is the access to the regular SPOT market)

            LD_API_KEY      = os.getenv("BINANCE_APIKEY")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")

            conn_exchange = ccxt.binance({
                'apiKey'         : LD_API_KEY,
                'secret'         : LD_API_SECRET,
                'enableRateLimit': True,
                'urls': {
                    'api': {
                        'Public': 'https://api.binance.com/api/v3/',
                        'Private':'https://api.binance.com/api/v3/',
                        }
                    },
            })
This Works (of course this is access to the regular SPOT market)


When I use the following:

            LD_API_KEY      = os.getenv("BINANCE_APIKEY")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")

            conn_exchange = ccxt.binance({
                'apiKey': LD_API_KEY,
                'secret': LD_API_SECRET,
                'timeout': 30000,
                'enableRateLimit': True,
                'options': { 'defaultType': 'future', },
            })

I get the API KEY error (again, this is the same KEY that works for the SPOT market). Also, I did hit the option for Futures.

When I use the following here:
```
            LD_API_KEY      = os.getenv("BINANCE_APIKEY")
            LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")
```

            conn_exchange = ccxt.binance({
                'apiKey': LD_API_KEY,
                'secret': LD_API_SECRET,
                'timeout': 30000,
                'enableRateLimit': True,
                'options': { 'defaultType': 'future', },
                'urls': {
                    'api': {
                        'fapiPublic': 'https://api.binance.com/fapi/v2',
                        'fapiPrivate':'https://api.binance.com/fapi/v2',
                        }
                    },
            })

I get:
error GET https://api.binance.com/fapi/v2/exchangeInfo 404 Not Found <html>

so ... this is not a connection error - but - it is a 404 error. I seem to have been able to get past the connection issue here(?)


I had wondered if it had something to do with the server being used:

I have tried these as options:
```
                'urls': {
                    'api': {
                        'fapiPublic': 'https://binancefuture.com/fapi/v1',
                        'fapiPrivate':'https://binancefuture.com/fapi/v1',
                        }
                    },
```
	---------------------------------------------------------------------------

```
                'urls': {
                     'api': {
                         'fapiPublic': 'https://api.binance.com/fapi/v2',
                         'fapiPrivate':'https://api.binance.com/fapi/v2',
                         }
                     },
```

	----------------------------------------------------------------------------
```
	I have also tried:
            #            'fapiPublic': 'https://api.binance.com/fapi/v1',
            #            'fapiPrivate':'https://api.binance.com/fapi/v1',
	[ ... and ... ]
            #            'public':  'https://fapi.binance.com/fapi/v1',
            #            'private': 'https://fapi.binance.com/fapi/v1',

```

Is there some other server(s) that I can try?

**Comment by cadmufasa** [comment]
Hello again, I also tried the following without success:

```
                'urls': {
                    'api': {
                        'Public': 'https://api.binance.com/fapi/v2',
                        'Private':'https://api.binance.com/fapi/v2',
                        }
                    },
```

**Comment by kroitor** [comment]
@cadmufasa you shouldn't override the urls this way. Have you checked the point (7) from my previous response?
Just use this:

```Python
        LD_API_KEY      = os.getenv("BINANCE_APIKEY")
        LD_API_SECRET   = os.getenv("BINANCE_API_SECRET")

        conn_exchange = ccxt.binance({
            'apiKey': LD_API_KEY,
            'secret': LD_API_SECRET,
            'timeout': 30000,
            'enableRateLimit': True,
            'options': { 'defaultType': 'future', },
        })
```

Have you enabled the futures trading mode upon creating the API key? 

<img width="1101" alt="Screen Shot 2020-07-13 at 19 09 39" src="https://user-images.githubusercontent.com/1294454/87327309-9662fa00-c53c-11ea-9a87-ead10bcc9b17.png">

Also, have you tried placing futures-orders via website? You have to pass a test there to enable the futures trading for your account.

**Comment by cadmufasa** [comment]
Thanks again for the response. I am working with your suggestions.
Regards.

**Comment by kroitor** [comment]
@cadmufasa let us know if you get it resolved. Thx!

**Comment by cadmufasa** [comment]
Hi kroitor,

Sometimes I feel like in some kind of alternate reality. I thought the "enable Margin" box I ticked said "enable Futures". I swear I did not see "enable Futures".

When you posted your picture, I looked at it again.

Also, i found that I had to create a new API Key any way (since the older API key was created before I turned on the Futures Account) 

When I did as you specified, all worked OK.

Thanks again.

Regards.
